src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(T value) {
+            this.value = value;
+        }
+
+        public Node(T value, Node<T> next, Node<T> previous) {
+            this.value = value;
+            this.next = next;
+            this.previous = previous;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (first == null) {
+            first = last = newNode;
+        } else {
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> newNode = new Node<>(value);
+        if (first == null) {
+            first = last = newNode;
+        } else if (index == 0) {
+            newNode.next = first;
+            first.previous = newNode;
+            first = newNode;
+        } else if (index == size) {
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        } else {
+            Node<T> preceding = getNodeByIndex(index - 1);
+            newNode.next = preceding.next;
+            newNode.next.previous = newNode;
+            newNode.previous = preceding;
+            preceding.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldValue;
+        checkIndex(index, size);
+        Node<T> node = getNodeByIndex(index);
+        oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        T removedValue;
+        Node<T> removedNode;
+        if (index == 0) {
+            removedValue = first.value;
+            removedNode = first;
+            first = first.next;
+            if (first == null) {
+                last = null;
+            }
+        } else {
+            Node<T> preceding = getNodeByIndex(index - 1);
+            removedNode = preceding.next;
+            removedValue = preceding.next.value;
+            preceding.next = preceding.next.next;
+            if (index == size - 1) {
+                last = preceding;
+            } else {
+                preceding.next.previous = preceding;
+            }
+        }
+        unlink(removedNode);
+        size--;
+        return removedValue;
```

Comment:
```suggestion
        checkIndex(index, size);
        T removedValue;
        Node<T> removedNode;
        if (index == 0) {
            removedValue = first.value;
            removedNode = first;
            first = first.next;
            if (first == null) {
                last = null;
            }
        } else {
            Node<T> preceding = getNodeByIndex(index - 1);
            removedNode = preceding.next;
            removedValue = preceding.next.value;
            preceding.next = preceding.next.next;
            if (index == size - 1) {
                last = preceding;
            } else {
                preceding.next.previous = preceding;
            }
        }
        unlink(removedNode);
        size--;
        return removedValue;
```

please reconsider of implementation of unlink method(). As written in checklist, we need to create all logic in unlink method() not in remove methods. (if you are not agree then later mentor can check or maybe ask mentor about this:) )))

Reply:
my implementation of the unlink() method implies deleting the links that were not deleted in remove() methods

Reply:
okay, i understand)